core/container/mutexMap_test.go

```diff
@@ -0,0 +1,265 @@
+
+	key := "key"
+	val1 := 56
+	val2 := 78
+
+	//operation succeed
+	ok := mm.Insert(key, val1)
+	assert.True(t, ok)
+
+	//test inner value
+	recoveredValue, _ := mm.Get(key)
+	assert.Equal(t, val1, recoveredValue)
+
+	//rewrite will fail
+	ok = mm.Insert(key, val2)
+	assert.False(t, ok)
+
+	//test inner value
+	recoveredValue, _ = mm.Get(key)
+	assert.Equal(t, val1, recoveredValue)
+}
+
+func TestMutexMap_InsertConcurrent(t *testing.T) {
+	t.Parallel()
+
+	mm := NewMutexMap()
+
+	numIterations := 100
+	wg := &sync.WaitGroup{}
+	wg.Add(numIterations)
+	for i := 0; i < numIterations; i++ {
+		go func(idx int) {
+			key := fmt.Sprintf("%d", idx)
+			ok := mm.Insert(key, struct{}{})
+			assert.True(t, ok)
+
+			wg.Done()
+		}(i)
+	}
+
+	wg.Wait()
+}
+
+//------- Remove
+
+func TestMutexMap_Remove(t *testing.T) {
+	t.Parallel()
+
+	mm := NewMutexMap()
+
+	key := "key"
+	mm.Remove(key)
+
+	mm.Set(key, struct{}{})
+	valRecovered, _ := mm.Get(key)
+	assert.NotNil(t, valRecovered)
+
+	mm.Remove(key)
+
+	valRecovered, _ = mm.Get(key)
+	assert.Nil(t, valRecovered)
+}
+
+func TestMutexMap_RemoveConcurrentShouldNotPanic(t *testing.T) {
+	t.Parallel()
+
+	defer func() {
+		r := recover()
+		if r != nil {
+			assert.Fail(t, fmt.Sprintf("should not have paniced %v", r))
+		}
+	}()
+
+	mm := NewMutexMap()
+
+	numIterations := 100
+	wg := &sync.WaitGroup{}
+	wg.Add(numIterations)
+	for i := 0; i < numIterations; i++ {
+		go func(idx int) {
+			mm.Remove("key")
+
+			wg.Done()
+		}(i)
+	}
+
+	wg.Wait()
+}
+
+//------- Keys
+
+func TestMutexMap_Keys(t *testing.T) {
+	t.Parallel()
+
+	mm := NewMutexMap()
+
+	key1 := "key1"
+	key2 := "key2"
+	key3 := "key3"
+
+	keys := mm.Keys()
+	assert.Equal(t, 0, len(keys))
+
+	mm.Set(key1, struct{}{})
+	mm.Set(key2, struct{}{})
+	mm.Set(key3, struct{}{})
+
+	keys = mm.Keys()
+	require.Equal(t, 3, len(keys))
+	stringKeys := make([]string, len(keys))
+	for idx, k := range keys {
+		stringKeys[idx] = k.(string)
+	}
+
+	sort.Slice(stringKeys, func(i, j int) bool {
+		return strings.Compare(stringKeys[i], stringKeys[j]) < 0
+	})
+
+	assert.Equal(t, key1, stringKeys[0])
+	assert.Equal(t, key2, stringKeys[1])
+	assert.Equal(t, key3, stringKeys[2])
+}
+
+func TestMutexMap_KeysConcurrent(t *testing.T) {
+	t.Parallel()
+
+	mm := NewMutexMap()
+	mm.Set("key1", struct{}{})
+	mm.Set("key2", struct{}{})
+	mm.Set("key3", struct{}{})
+
+	numIterations := 100
+	wg := &sync.WaitGroup{}
+	wg.Add(numIterations)
+	for i := 0; i < numIterations; i++ {
+		go func(idx int) {
+			keys := mm.Keys()
+			assert.Equal(t, 3, len(keys))
+
+			wg.Done()
+		}(i)
+	}
+
+	wg.Wait()
+}
+
+//------- Len
+
+func TestMutexMap_Len(t *testing.T) {
+	t.Parallel()
+
+	mm := NewMutexMap()
+
+	key1 := "key1"
+	key2 := "key2"
+	key3 := "key3"
+
+	assert.Equal(t, 0, mm.Len())
+
+	mm.Set(key1, struct{}{})
+	assert.Equal(t, 1, mm.Len())
+
+	mm.Set(key2, struct{}{})
+	assert.Equal(t, 2, mm.Len())
+
+	mm.Set(key3, struct{}{})
+	assert.Equal(t, 3, mm.Len())
+}
```

Comment:
There are no tests with concurrent removals & getters & setters - paired or all of them running concurrently. I see `t.Parallel()` though, so `go test -race` should cover those.